### PR TITLE
(un)pickling support for ADTs with associated, user-defined values (closes #28)

### LINKED ISSUE #28: Add ordering for case enum
case enum would be much more useful if there was a way to convert to and from Int, based on the ordering. This would give significant space savings in serialization protocols

### PR DESCRIPTION
**Pull request by utaal**
``` scala
@indexedenum trait Planet {
  type Index = Int
  object Mercury { 0 }
  object Venus   { 3 }
}
```

The `Int => Planet` conversion is provided via a typeclass:

``` scala
scala> Planet.Mercury.index
res0: 0: Int
scala> Planet.caseFromIndex(3)
res1: Some(Planet.Venus): Option[Planet]
```
## Test plan

Unit tests.

### CODE REVIEW

**Comment by calippo** [comment]
Going in vacation for a few days, re-assigning to @danielegallingani

**Comment by gabro** [comment]
`indexedenum`, really?
Why not `askdjhakiq`?

File: caseenum/src/main/scala/macro.scala [hunk 1]
```diff
@@ -63,7 +63,81 @@ object EnumMacro {
...
+}
+
+/**
+ * Macro annotation that transforms a trait enclosing a single type
+ * alias and a set of objects to an ADT following the IndexedCaseEnum
+ * convention (see the IndexedCaseEnum trait).
+ *
+ * e.g.
+ * ```
+ * @indexedenum trait Planet(Int) {
```
Comment (gabro) [inline caseenum/src/main/scala/macro.scala]: `Planet(Int)`?

**Comment by gabro** [comment]
Looks good, although the documentation is slightly misaligned with the actual implementation and I would rename it `indexedEnum`

File: caseenum/src/main/scala/macro.scala (hunk 1 above)
Comment (utaal) [inline caseenum/src/main/scala/macro.scala]: Yup, old (unimplementable) syntax. Good catch.

**Comment by utaal** [comment]
Have you noticed other documentation issues besides the one you commented on? I thought I caught them all ;)

**Pull request merged**